third_party/amd/lib/TritonAMDGPUToLLVM/TargetInfo.cpp

```diff
@@ -224,12 +224,101 @@ static inline Value truncAndCastFromInt(RewriterBase &rewriter, Location loc,
+  Type retType = struct_ty({i32_ty, i32_ty});
+  Value perm =
+      LLVM::createLLVMIntrinsicCallOp(rewriter, loc, intrinsic, retType,
+                                      ValueRange{val, val, falseVal, falseVal})
+          ->getResult(0);
+  Value v0 = b.extract_val(i32_ty, perm, 0);
+  Value v1 = b.extract_val(i32_ty, perm, 1);
+
+  if (!valType.isInteger(32)) {
+    v0 = truncAndCastFromInt(rewriter, loc, v0, valType, 32);
+    v1 = truncAndCastFromInt(rewriter, loc, v1, valType, 32);
+  }
+  IRMapping mapping;
+  mapping.map(reduxOp->getOperand(0), v0);
+  mapping.map(reduxOp->getOperand(1), v1);
+  Value redx = rewriter.clone(*reduxOp, mapping)->getResult(0);
+  return redx;
+}
+
+// Apply warp reduction across lanes using llvm intrinsics in GFX950.
+// The input acc has the partial accumulated value from reduction within
+// threads. The output acc has the final accumulated value.
+//
+// The thread layout must be mfma input layout.
+// For the mfma_32x32 layout:
+//   step 1: use permlane32_swap() to swap the row 2 and 3 of acc and
+//           the row 0 and 1 of the copy of acc
+//   step 2: apply reduction to the result values to get final result
+// For the mfma_16x16 layout:
+//   step 1: use permlane32_swap() to swap the row 2 and 3 of acc and
+//           the row 0 and 1 of the copy of acc
+//   step 2: apply reduction to the result values to get the partial result
+//   step 3: use permlane16_swap() to swap the odd and even rows of
+//           the partial results
+//   step 4: apply reduction to get the final results
+static bool warpReduceCDNA4(RewriterBase &rewriter, Location loc,
+                            SmallVector<Value> &acc, triton::ReduceOp op) {
+  Operation *reduxOp = op.getSingleCombiner();
+  if (!reduxOp)
+    return false;
+
+  if (acc.size() != 1)
+    return false;
+  Value val = acc[0];
+  unsigned bits = val.getType().getIntOrFloatBitWidth();
+  if (bits > 32)
+    return false;
+
+  auto resultTy = op.getInputTypes()[0];
+  auto mfmaLayout = dyn_cast<AMDMfmaEncodingAttr>(resultTy.getEncoding());
+  if (!mfmaLayout)
+    return false;
+
+  auto MDim = mfmaLayout.getMDim();
+  auto NDim = mfmaLayout.getNDim();
+  StringRef intrinsic;
+  if (MDim == 32 && NDim == 32 || MDim == 16 && NDim == 16)
+    intrinsic = "llvm.amdgcn.permlane32.swap";
+  else
+    return false;
+
+  Value redx = permuteAndReduce(rewriter, loc, intrinsic, val, reduxOp);
+
+  if (MDim == 16 && NDim == 16) {
+    intrinsic = "llvm.amdgcn.permlane16.swap";
+    redx = permuteAndReduce(rewriter, loc, intrinsic, redx, reduxOp);
+  }
+
+  acc[0] = redx;
+  return true;
+}
+
 bool TargetInfo::warpReduce(RewriterBase &rewriter, Location loc,
                             SmallVector<Value> &acc, triton::ReduceOp op,
                             unsigned numLaneToReduce,
                             unsigned interleave) const {
   auto b = TritonLLVMOpBuilder(loc, rewriter);
 
+  if (isCDNA() && getISAFamily() == ISAFamily::CDNA4 &&
```

Comment:
You want to put this check below the check for `numLaneToReduce` so that it can early return.

Reply:
> You want to put this check below the check for `numLaneToReduce` so that it can early return.

For the FA test case, numLaneToReduce (=2) is not equal to warpSize (=64), so it would exit without reaching warpReduceCDNA4.

Reply:
So previously we expect TargetInfo::warpReduce to handle reduction of 64 threads?

Ok. I check the logic of warpReduce in the common codebase. Here we are handling a special case of warpReduce where
- For mfma32
    - `numLaneToReduce` = 2
    - `interleave` = 32
- For mfma16
    - `numLaneToReduce` = 4
    - `interleave` = 16 

Right?